File: rust/fuzz/fuzz_targets/verify-transaction.rs
#![no_main]

//! The StoreKit 2 path: compact-JWS split, strict base64url, JSON header
//! and payload, `x5c` certificates, chain, ES256 signature, then the three
//! public entry points' claim checks. Same invariants as the Go port's
//! `FuzzVerifyTransaction`: nothing panics, and a JWS `verify_raw` accepts
//! under the fixture root must be refused under Apple's roots, or the
//! anchors are not what decided it.

use apple_purchase_receipt_verifier::{apple_jws_roots, Environment, JwsVerifier, TrustAnchor};
use libfuzzer_sys::fuzz_target;
use std::sync::OnceLock;

const JWS_ROOT: &[u8] = include_bytes!("../../../fixtures/generated/jws-root.der");

fn verifiers() -> &'static (JwsVerifier, JwsVerifier) {
    static VERIFIERS: OnceLock<(JwsVerifier, JwsVerifier)> = OnceLock::new();
    VERIFIERS.get_or_init(|| {
        let fixture = JwsVerifier::builder()
            .trusted_roots([TrustAnchor::from_der(JWS_ROOT).expect("fixture root")])
            .bundle_id("com.example.app")
            .accepted_environments([Environment::Sandbox])
            .build()
            .expect("a non-empty anchor set and a bundle id");
        let unrelated = JwsVerifier::builder()
            .trusted_roots(apple_jws_roots().iter().cloned())
            .bundle_id("com.example.app")
            .accepted_environments([Environment::Sandbox])
            .build()
            .expect("a non-empty anchor set and a bundle id");
        (fixture, unrelated)
    })
}

fuzz_target!(|data: &[u8]| {
    let Ok(jws) = std::str::from_utf8(data) else {
        return;
    };
    let (verifier, unrelated) = verifiers();
    let _ = verifier.verify_transaction(jws);
    let _ = verifier.verify_app_transaction(jws);
    if verifier.verify_raw(jws).is_ok() {
        assert!(
            unrelated.verify_raw(jws).is_err(),
            "this input verifies against Apple's roots too, \
             so the anchors are not being enforced"
        );
    }
});

File: php/fuzz/targets/verify-transaction.php
<?php

declare(strict_types=1);

namespace EminDeniz99\ApplePurchaseReceiptVerifier\Fuzz;

use EminDeniz99\ApplePurchaseReceiptVerifier\AppleRootCerts;
use EminDeniz99\ApplePurchaseReceiptVerifier\Environment;
use EminDeniz99\ApplePurchaseReceiptVerifier\Jws\JwsVerifier;
use EminDeniz99\ApplePurchaseReceiptVerifier\VerificationException;

/**
 * The StoreKit 2 path: compact-JWS split, strict base64url, JSON header and
 * payload, `x5c` certificates, chain, ES256 signature, then the three public
 * entry points' claim checks.
 *
 * Same invariants as the Go port's `FuzzVerifyTransaction`: nothing escapes
 * but `VerificationException`, and a JWS `verifyRaw` accepts under the
 * fixture root must be refused under Apple's roots, or the anchors are not
 * what decided it.
 */

/** @var \PhpFuzzer\Config $config */
require __DIR__ . '/../bootstrap.php';

$verifier = new JwsVerifier(
    FuzzFixtures::jwsRootOnly(),
    'com.example.app',
    [Environment::Sandbox],
);
$unrelated = new JwsVerifier(
    AppleRootCerts::jwsRoots(),
    'com.example.app',
    [Environment::Sandbox],
);

$config->setMaxLen(8192);
$config->setAllowedExceptions([VerificationException::class]);

$config->setTarget(static function (string $input) use ($verifier, $unrelated): void {
    // Each entry point is exercised even when an earlier one rejects: a
    // VerificationException is a verdict, and only something else is a
    // finding. Anything that is not a VerificationException propagates out of
    // these blocks and the fuzzer records it as a crash.
    try {
        $verifier->verifyTransaction($input);
    } catch (VerificationException) {
    }
    try {
        $verifier->verifyAppTransaction($input);
    } catch (VerificationException) {
    }

    try {
        $verifier->verifyRaw($input);
    } catch (VerificationException) {
        return;
    }

    try {
        $unrelated->verifyRaw($input);
    } catch (VerificationException) {
        return;
    }

    throw new \Error(
        'this input verifies against Apple\'s roots too, '
        . 'so the anchors are not being enforced',
    );
});
